quota-3.14/setquota.8
.TH SETQUOTA 8
.SH NAME
setquota \- set disk quotas
.SH SYNOPSIS
.B setquota
[
.B \-r
]
[
.B \-u
|
.B \-g
]
[
.B \-F
.I quotaformat
]
.I name
.I block-softlimit
.I block-hardlimit
.I inode-softlimit
.I inode-hardlimit
.B \-a
|
.I filesystem...
.LP
.B setquota
[
.B \-r
]
[
.B \-u
|
.B \-g
]
[
.B \-F
.I quotaformat
]
[
.B \-p
.I protoname 
]
.I name
.B \-a
|
.I filesystem...
.LP
.B setquota
.B \-b
[
.B \-r
]
[
.B \-u
|
.B \-g
]
[
.B \-F
.I quotaformat
]
.B \-a
|
.I filesystem...
.LP
.B setquota
.B \-t
[
.B \-u
|
.B \-g
]
[
.B \-F
.I quotaformat
]
.I block-grace
.I inode-grace
.B \-a
|
.I filesystem...
.LP
.B setquota
.B \-T
[
.B \-u
|
.B \-g
]
[
.B \-F
.I quotaformat
]
.I name
.I block-grace
.I inode-grace
.B \-a
|
.I filesystem...
.SH DESCRIPTION
.IX  "setquota command"  ""  "\fLsetquota\fP \(em set disk quotas"
.IX  set "disk quotas \(em \fLsetquota\fP"
.IX  "disk quotas"  "setquota command"  ""  "\fLsetquota\fP \(em set disk quotas"
.IX  "disk quotas"  "setquota command"  ""  "\fLsetquota\fP \(em set disk quotas"
.IX  "quotas"  "setquota command"  ""  "\fLsetquota\fP \(em set disk quotas"
.IX  "filesystem"  "setquota command"  ""  "\fLsetquota\fP \(em set disk quotas"
.B setquota
is a command line quota editor.
The filesystem, user/group name and new quotas for this
filesystem can be specified on the command line. Note that if a number is
given in the place of a user/group name it is treated as an UID/GID.
.TP
.B -r, --remote
Edit also remote quota use rpc.rquotad on remote server to set quota. This
option is available only if quota tools were compiled with enabled support
for setting quotas over RPC.
.TP
.B -F, --format=\f2quotaformat\f1
Perform setting for specified format (ie. don't perform format autodetection).
Possible format names are:
.B vfsold
(version 1 quota),
.B vfsv0
(version 2 quota),
.B rpc
(quota over NFS),
.B xfs
(quota on XFS filesystem)
.TP
.B -u, --user
Set user quotas for named user. This is the default.
.TP
.B -g, --group
Set group quotas for named group.
.TP
.B -p, --prototype=\f2protoname\f1
Use quota settings of user or group
.I protoname
to set the quota for the named user or group.
.TP
.B -b, --batch
Read information to set from stdin (input format is
.I name block-softlimit block-hardlimit inode-softlimit inode-hardlimit
).
.TP
.B -t, --edit-period
Set grace times for users/groups. Times
.B block-grace
and
.B inode-grace
are specified in seconds.
.TP
.B -T, --edit-times
Alter times for individual user/group when softlimit is enforced. Times
.B block-grace
and
.B inode-grace
are specified in seconds or can be string 'unset'.
.TP
.B -a, --all
Go through all filesystems with quota in
.B /etc/mtab
and perform setting.
.PP
To disable a quota, set the coresponding parameter to 0. To change quotas
for several filesystems, invoke once for each filesystem.
.PP
Only the super-user may edit quotas.
.SH FILES
.PD 0
.TP 20
.B aquota.user or aquota.group
quota file at the filesystem root (version 2 quota, non-XFS filesystems)
.TP
.B quota.user or quota.group
quota file at the filesystem root (version 1 quota, non-XFS filesystems)
.TP
.B /etc/mtab
mounted filesystem table
.PD
.SH SEE ALSO
.BR edquota (8),
.BR quota (1),
.BR quotactl (2),
.BR quotacheck (8),
.BR quotaon (8),
.BR repquota (8)


noflushd/NEWS
* noflushd 2.7.5 (09. May 2005)

  - Fix handling of raid and similar meta devices on 2.6 kernels.

* noflushd 2.7.4 (08. Aug 2004)

  - Fix lousy spindown times on 2.6 kernels.
  - Fix SCSI support on 2.6 kernels.
  - Fix pid detection in generic startup script.
  - Document quirks with hotplugged drives.

* noflushd 2.7.3 (09. Apr 2004)

  - Fix race condition that might leave kupdated/pdflushd disabled
    aftern noflushd has terminated.
  - Fix spindown when all disks are auto-detected.

* noflushd 2.7.2 (16. Mar 2004)

  - A change in 2.7.1 erroneously flushed the buffer cache before
    spindown, causing frequent spinups. This change has been reverted.
    
* noflushd 2.7.1 (07. Mar 2004)

  - Fix immediate spinup on 2.6 kernels.
  - Elevate SCSI support from compile-time to command line option.
  - Bug fixes in the disk statistics parser for 2.6 kernels.
  - Ship kernel patch to track down which processes cause spinups.

* noflushd 2.7 (20. Jan 2004)

  - Support Linux kernel 2.6.x.
  - Default to monitorring all disks unless specific disks are given
    on the command line. This change should yield a sane default
    behaviour for a majority of users without requiring explicit
    configuration.

* noflushd 2.6.3 (13. May 2002)

  - Bugfix for arm and other archs with weird /proc/partitions layout.

* noflushd 2.6.2 (20. Apr 2002)

  - Add IDE raid support.
  - Add timeout logging script.

* noflushd 2.6.1 (03. Feb. 2002)

  - Fix BUG when using noflushd with disks on second IDE channel on 2.4
    kernels.

* noflushd 2.6 (22. Jan 2002)

  - Document noflushd's suckage with any journaling fs, including ext3.
  - New feature prevents spindown if machine is in interactive use.
  - Add hack fix some '0 spindown time' uglies.

* noflushd 2.5 (06. Aug 2001)

  - Make sure hotplugged disks get synced properly.
  - Only sync present partitions. (Fixes high CPU usage bug on devfs
    systems.)
  - Generic startup skript works on latest 2.4 kernels.

* noflushd 2.4 (13. Mar 2001)

  - Timeout lists are back to a working state.
  - Support very latest 2.4 kernels.

* noflushd 2.3 (12. Jan 2001)

  - Support latest 2.4 kernels. (Leaves kernels 2.3.99pre8 to 2.4.0-test8
    unsupported.)
  - Fix multi-disk support with long bdflush timeouts.

* noflushd 2.2 (22. Dec 2000)

  - Fixed bug in spindown code. Now actually spins down. Duh!

* noflushd 2.1 (20. Dec 2000)

  - Disks get synced prior to spindown.
  - Improved devfs support.

* noflushd 2.0 (14. Nov 2000)

  - noflushd now comes with its own syncer that takes over the task from
    kupdate when the first read-write mounted disk is spun down. This
    enhances spindown times in a multidisk environment at the expense of
    less smooth and more bursty background flushing when the noflushd 
    syncer is in use.

  - Conflicting entries in 2.2-style /proc/stat (e.g. /dev/sda and /dev/hda)
    won't cause noflushd to give up anymore. Instead, noflushd will treat
    both disks pretty much like one, i.e. they're either both up or spun
    down, no matter if activity only goes to one of them.

  - Compilation environment switched to automake/autoconf.

  - When installing via 'make install', noflushd is no longer added to
    any runlevel. 

  - noflushd moved to SourceForge. The project homepage is located at
    http://noflushd.sourceforge.net, a mailing list is available via
    noflushd-devel@lists.sourceforge.net. Visit the homepage for 
    subscription information.



quota-4.00~pre1/setquota.c
/*
 *
 *	Set disk quota from command line 
 *
 *	Jan Kara <jack@suse.cz> - sponsored by SuSE CR
 */

#include "config.h"

#include <rpc/rpc.h>
#include <sys/types.h>
#include <errno.h>
#include <stdio.h>
#include <string.h>
#include <getopt.h>
#include <time.h>
#include <ctype.h>

#if defined(RPC)
#include "rquota.h"
#include "rquota_client.h"
#endif
#include "pot.h"
#include "quotaops.h"
#include "common.h"
#include "quotasys.h"

#define FL_USER 1
#define FL_GROUP 2
#define FL_RPC 4
#define FL_ALL 8
#define FL_PROTO 16
#define FL_GRACE 32
#define FL_INDIVIDUAL_GRACE 64
#define FL_BATCH 128
#define FL_NUMNAMES 256
#define FL_NO_MIXED_PATHS 512
#define FL_CONTINUE_BATCH 1024

int flags, fmt = -1;
char **mnt;
char *progname;
int mntcnt;
qid_t protoid, id;
struct util_dqblk toset;

/* Print usage information */
static void usage(void)
{
#if defined(RPC_SETQUOTA)
	char *ropt = "[-rm] ";
#else
	char *ropt = "";
#endif
	errstr(_("Usage:\n\
  setquota [-u|-g] %1$s[-F quotaformat] <user|group>\n\
\t<block-softlimit> <block-hardlimit> <inode-softlimit> <inode-hardlimit> -a|<filesystem>...\n\
  setquota [-u|-g] %1$s[-F quotaformat] <-p protouser|protogroup> <user|group> -a|<filesystem>...\n\
  setquota [-u|-g] %1$s[-F quotaformat] -b [-c] -a|<filesystem>...\n\
  setquota [-u|-g] [-F quotaformat] -t <blockgrace> <inodegrace> -a|<filesystem>...\n\
  setquota [-u|-g] [-F quotaformat] <user|group> -T <blockgrace> <inodegrace> -a|<filesystem>...\n\n\
-u, --user                 set limits for user\n\
-g, --group                set limits for group\n\
-a, --all                  set limits for all filesystems\n\
    --always-resolve       always try to resolve name, even if is\n\
                           composed only of digits\n\
-F, --format=formatname    operate on specific quota format\n\
-p, --prototype=protoname  copy limits from user/group\n\
-b, --batch                read limits from standard input\n\
-c, --continue-batch       continue in input processing in case of an error\n"), ropt);
#if defined(RPC_SETQUOTA)
	fputs(_("-r, --remote               set remote quota (via RPC)\n\
-m, --no-mixed-pathnames      trim leading slashes from NFSv4 mountpoints\n"), stderr);
#endif
	fputs(_("-t, --edit-period          edit grace period\n\
-T, --edit-times           edit grace times for user/group\n\
-h, --help                 display this help text and exit\n\
-V, --version              display version information and exit\n\n"), stderr);
	fprintf(stderr, _("Bugs to: %s\n"), MY_EMAIL);
	exit(1);
}

/* Convert string to number - print errstr message in case of failure */
static qsize_t parse_unum(char *str, char *msg)
{
	char *errch;
	qsize_t ret = strtoull(str, &errch, 0);

	if (*errch) {
		errstr(_("Bad %s: %s\n"), msg, str);
		usage();
	}
	return ret;
}

/* Convert our flags to quota type */
static inline int flag2type(int flags)
{
	if (flags & FL_USER)
		return USRQUOTA;
	if (flags & FL_GROUP)
		return GRPQUOTA;
	return -1;
}

/* Parse options of setquota */
static void parse_options(int argcnt, char **argstr)
{
	int ret, otherargs;
	char *protoname = NULL;

#ifdef RPC_SETQUOTA
	char *opts = "gp:urmVF:taTbc";
#else
	char *opts = "gp:uVF:taTbc";
#endif
	struct option long_opts[] = {
		{ "user", 0, NULL, 'u' },
		{ "group", 0, NULL, 'g' },
		{ "prototype", 1, NULL, 'p' },
#ifdef RPC_SETQUOTA
		{ "remote", 0, NULL, 'r' },
		{ "no-mixed-pathnames", 0, NULL, 'm' },
#endif
		{ "all", 0, NULL, 'a' },
		{ "always-resolve", 0, NULL, 256},
		{ "edit-period", 0, NULL, 't' },
		{ "edit-times", 0, NULL, 'T' },
		{ "batch", 0, NULL, 'b' },
		{ "continue", 0, NULL, 'c' },
		{ "format", 1, NULL, 'F' },
		{ "version", 0, NULL, 'V' },
		{ "help", 0, NULL, 'h' },
		{ NULL, 0, NULL, 0 }
	};

	while ((ret = getopt_long(argcnt, argstr, opts, long_opts, NULL)) != -1) {
		switch (ret) {
		  case '?':
		  case 'h':
			  usage();
		  case 'g':
			  flags |= FL_GROUP;
			  break;
		  case 'u':
			  flags |= FL_USER;
			  break;
		  case 'p':
			  flags |= FL_PROTO;
			  protoname = optarg;
			  break;
		  case 'r':
			  flags |= FL_RPC;
			  break;
		  case 'm':
			  flags |= FL_NO_MIXED_PATHS;
			  break;
		  case 'a':
			  flags |= FL_ALL;
			  break;
		  case 256:
			  flags |= FL_NUMNAMES;
			  break;
		  case 't':
			  flags |= FL_GRACE;
			  break;
		  case 'b':
			  flags |= FL_BATCH;
			  break;
		  case 'c':
			  flags |= FL_CONTINUE_BATCH;
			  break;
		  case 'T':
			  flags |= FL_INDIVIDUAL_GRACE;
			  break;
		  case 'F':
			  if ((fmt = name2fmt(optarg)) == QF_ERROR)
				  exit(1);
			  break;
		  case 'V':
			  version();
			  exit(0);
		}
	}
	if (flags & FL_USER && flags & FL_GROUP) {
		errstr(_("Group and user quotas cannot be used together.\n"));
		usage();
	}
	if (flags & FL_PROTO && flags & FL_GRACE) {
		errstr(_("Prototype user has no sense when editing grace times.\n"));
		usage();
	}
	if (flags & FL_INDIVIDUAL_GRACE && flags & FL_GRACE) {
		errstr(_("Cannot set both individual and global grace time.\n"));
		usage();
	}
	if (flags & FL_BATCH && flags & (FL_GRACE | FL_INDIVIDUAL_GRACE)) {
		errstr(_("Batch mode cannot be used for setting grace times.\n"));
		usage();
	}
	if (flags & FL_BATCH && flags & FL_PROTO) {
		errstr(_("Batch mode and prototype user cannot be used together.\n"));
		usage();
	}
	if (flags & FL_GRACE)
		otherargs = 2;
	else if (flags & FL_INDIVIDUAL_GRACE)
		otherargs = 3;
	else if (flags & FL_BATCH)
		otherargs = 0;
	else {
		otherargs = 1;
		if (!(flags & FL_PROTO))
			otherargs += 4;
	}
	if (optind + otherargs > argcnt) {
		errstr(_("Bad number of arguments.\n"));
		usage();
	}
	if (!(flags & (FL_USER | FL_GROUP)))
		flags |= FL_USER;
	if (!(flags & (FL_GRACE | FL_BATCH))) {
		id = name2id(argstr[optind++], flag2type(flags), !!(flags & FL_NUMNAMES), NULL);
		if (!(flags & (FL_GRACE | FL_INDIVIDUAL_GRACE | FL_PROTO))) {
			toset.dqb_bsoftlimit = parse_unum(argstr[optind++], _("block softlimit"));
			toset.dqb_bhardlimit = parse_unum(argstr[optind++], _("block hardlimit"));
			toset.dqb_isoftlimit = parse_unum(argstr[optind++], _("inode softlimit"));
			toset.dqb_ihardlimit = parse_unum(argstr[optind++], _("inode hardlimit"));
		}
		else if (flags & FL_PROTO)
			protoid = name2id(protoname, flag2type(flags), !!(flags & FL_NUMNAMES), NULL);
	}
	if (flags & FL_GRACE) {
		toset.dqb_btime = parse_unum(argstr[optind++], _("block grace time"));
		toset.dqb_itime = parse_unum(argstr[optind++], _("inode grace time"));
	}
	else if (flags & FL_INDIVIDUAL_GRACE) {
		time_t now;

		time(&now);
		if (!strcmp(argstr[optind], _("unset"))) {
			toset.dqb_btime = 0;
			optind++;
		}
		else
			toset.dqb_btime = now + parse_unum(argstr[optind++], _("block grace time"));
		if (!strcmp(argstr[optind], _("unset"))) {
			toset.dqb_itime = 0;
			optind++;
		}
		else
			toset.dqb_itime = now + parse_unum(argstr[optind++], _("inode grace time"));
	}
	if (!(flags & FL_ALL)) {
		mntcnt = argcnt - optind;
		mnt = argstr + optind;
		if (!mntcnt) {
			errstr(_("Mountpoint not specified.\n"));
			usage();
		}
	}
}

/* Set user limits */
static int setlimits(struct quota_handle **handles)
{
	struct dquot *q, *protoq, *protoprivs = NULL, *curprivs;
	int ret = 0;

	curprivs = getprivs(id, handles, 0);
	if (flags & FL_PROTO) {
		protoprivs = getprivs(protoid, handles, 0);
		for (q = curprivs, protoq = protoprivs; q && protoq; q = q->dq_next, protoq = protoq->dq_next) {
			q->dq_dqb.dqb_bsoftlimit = protoq->dq_dqb.dqb_bsoftlimit;
			q->dq_dqb.dqb_bhardlimit = protoq->dq_dqb.dqb_bhardlimit;
			q->dq_dqb.dqb_isoftlimit = protoq->dq_dqb.dqb_isoftlimit;
			q->dq_dqb.dqb_ihardlimit = protoq->dq_dqb.dqb_ihardlimit;
			update_grace_times(q);
		}
		freeprivs(protoprivs);
	}
	else {
		for (q = curprivs; q; q = q->dq_next) {
			q->dq_dqb.dqb_bsoftlimit = toset.dqb_bsoftlimit;
			q->dq_dqb.dqb_bhardlimit = toset.dqb_bhardlimit;
			q->dq_dqb.dqb_isoftlimit = toset.dqb_isoftlimit;
			q->dq_dqb.dqb_ihardlimit = toset.dqb_ihardlimit;
			update_grace_times(q);
		}
	}
	if (putprivs(curprivs, COMMIT_LIMITS) == -1)
		ret = -1;
	freeprivs(curprivs);
	return ret;
}

#define MAXLINELEN 65536

/* Read & parse one batch entry */
static int read_entry(qid_t *id, qsize_t *isoftlimit, qsize_t *ihardlimit, qsize_t *bsoftlimit, qsize_t *bhardlimit)
{
	static int line = 0;
	char name[MAXNAMELEN+1];
	char linebuf[MAXLINELEN], *chptr;
	unsigned long is, ih, bs, bh;
	int ret;

	while (1) {
		line++;
		if (!fgets(linebuf, sizeof(linebuf), stdin))
			return -1;
		if (linebuf[strlen(linebuf)-1] != '\n')
			die(1, _("Line %d too long.\n"), line);
		/* Comment? */
		if (linebuf[0] == '#')
			continue;
		/* Blank line? */
		chptr = linebuf;
		while (isblank(*chptr))
			chptr++;
		if (*chptr == '\n')
			continue;
		ret = sscanf(chptr, "%s %lu %lu %lu %lu", name, &bs, &bh, &is, &ih);
		if (ret != 5) {
			errstr(_("Cannot parse input line %d.\n"), line);
			if (!(flags & FL_CONTINUE_BATCH))
				die(1, _("Exitting.\n"));
			errstr(_("Skipping line.\n"));
			continue;
		}
		*id = name2id(name, flag2type(flags), !!(flags & FL_NUMNAMES), &ret);
		if (ret) {
			errstr(_("Unable to resolve name '%s' on line %d.\n"), name, line);
			if (!(flags & FL_CONTINUE_BATCH))
				die(1, _("Exitting.\n"));
			errstr(_("Skipping line.\n"));
			continue;
		}
		break;
	}
	*isoftlimit = is;
	*ihardlimit = ih;
	*bsoftlimit = bs;
	*bhardlimit = bh;
	return 0;
}

/* Set user limits in batch mode */
static int batch_setlimits(struct quota_handle **handles)
{
	struct dquot *curprivs, *q;
	qsize_t bhardlimit, bsoftlimit, ihardlimit, isoftlimit;
	qid_t id;
	int ret = 0;

	while (!read_entry(&id, &isoftlimit, &ihardlimit, &bsoftlimit, &bhardlimit)) {
		curprivs = getprivs(id, handles, 0);
		for (q = curprivs; q; q = q->dq_next) {
			q->dq_dqb.dqb_bsoftlimit = bsoftlimit;
			q->dq_dqb.dqb_bhardlimit = bhardlimit;
			q->dq_dqb.dqb_isoftlimit = isoftlimit;
			q->dq_dqb.dqb_ihardlimit = ihardlimit;
			update_grace_times(q);
		}
		if (putprivs(curprivs, COMMIT_LIMITS) == -1)
			ret = -1;
		freeprivs(curprivs);
	}
	return ret;
}

/* Set grace times */
static int setgraces(struct quota_handle **handles)
{
	int i;

	for (i = 0; handles[i]; i++) {
		handles[i]->qh_info.dqi_bgrace = toset.dqb_btime;
		handles[i]->qh_info.dqi_igrace = toset.dqb_itime;
		mark_quotafile_info_dirty(handles[i]);
	}
	return 0;
}

/* Set grace times for individual user */
static int setindivgraces(struct quota_handle **handles)
{
	int ret = 0;
	struct dquot *q, *curprivs;

	curprivs = getprivs(id, handles, 0);
	for (q = curprivs; q; q = q->dq_next) {
		if (q->dq_dqb.dqb_bsoftlimit && toqb(q->dq_dqb.dqb_curspace) > q->dq_dqb.dqb_bsoftlimit)
			q->dq_dqb.dqb_btime = toset.dqb_btime;
		else
			errstr(_("Not setting block grace time on %s because softlimit is not exceeded.\n"), q->dq_h->qh_quotadev);
		if (q->dq_dqb.dqb_isoftlimit && q->dq_dqb.dqb_curinodes > q->dq_dqb.dqb_isoftlimit)
			q->dq_dqb.dqb_itime = toset.dqb_itime;
		else
			errstr(_("Not setting inode grace time on %s because softlimit is not exceeded.\n"), q->dq_h->qh_quotadev);
	}
	if (putprivs(curprivs, COMMIT_TIMES) == -1) {
		errstr(_("cannot write times for %s. Maybe kernel does not support such operation?\n"), type2name(flags & FL_USER ? USRQUOTA : GRPQUOTA));
		ret = -1;
	}
	freeprivs(curprivs);
	return ret;
}

int main(int argc, char **argv)
{
	struct quota_handle **handles;
	int ret;

	gettexton();
	progname = basename(argv[0]);

	parse_options(argc, argv);
	init_kernel_interface();

	if (flags & FL_ALL)
		handles = create_handle_list(0, NULL, flag2type(flags), fmt,
			(flags & FL_NO_MIXED_PATHS) ? 0 : IOI_NFS_MIXED_PATHS,
			(flags & FL_RPC) ? 0 : MS_LOCALONLY);
	else
		handles = create_handle_list(mntcnt, mnt, flag2type(flags), fmt,
			(flags & FL_NO_MIXED_PATHS) ? 0 : IOI_NFS_MIXED_PATHS,
			(flags & FL_RPC) ? 0 : MS_LOCALONLY);

	if (flags & FL_GRACE)
		ret = setgraces(handles);
	else if (flags & FL_INDIVIDUAL_GRACE)
		ret = setindivgraces(handles);
	else if (flags & FL_BATCH)
		ret = batch_setlimits(handles);
	else
		ret = setlimits(handles);

	if (dispose_handle_list(handles) == -1)
		ret = -1;

	return ret ? 1 : 0;
}
